### PR TITLE
Container soft delete

### PR DESCRIPTION
**Pull request by souravgupta-msft**
Added code for restoring soft-deleted container

- [x] The purpose of this PR is explained in this or a referenced issue.
- [x] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [x] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [x] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

File: sdk/storage/azblob/container/client.go [hunk 1]
```diff
@@ -160,6 +160,18 @@ func (c *Client) Delete(ctx context.Context, options *DeleteOptions) (DeleteResp
 	return resp, err
 }
 
+// Restore operation restore the contents and properties of a soft deleted container to a specified container.
+// For more information, see https://docs.microsoft.com/en-us/rest/api/storageservices/restore-container.
+func (c *Client) Restore(ctx context.Context, deletedContainerName string, deletedContainerVersion string) (RestoreResponse, error) {
```
Comment (jhendrixMSFT) [inline sdk/storage/azblob/container/client.go:165]: Our guidelines state that APIs should have an options type as the last argument.  This is to provide extensibility without a breaking change.

Also, are both of these args mandatory?  I see they're not in the generated code.  I know that sometimes the generated code is incorrect due to OpenAPI spec bugs.  Is that the case here?

File: sdk/storage/azblob/container/client.go [hunk 2]
```diff
@@ -160,6 +160,18 @@ func (c *Client) Delete(ctx context.Context, options *DeleteOptions) (DeleteResp
...
+	opts := &generated.ContainerClientRestoreOptions{
```
Comment (vibhansa-msft) [inline sdk/storage/azblob/container/client.go:166]: We shall redefine this in exported code and that can be used as input options to this API.

File: sdk/storage/azblob/container/client.go (hunk 1 above)
Comment (souravgupta-msft) [reply sdk/storage/azblob/container/client.go]: Converted the args to options

File: sdk/storage/azblob/container/client.go (hunk 2 above)
Comment (souravgupta-msft) [reply sdk/storage/azblob/container/client.go]: Done

**Review by tasherif-msft — APPROVED** [review]

**Review by siminsavani-msft — APPROVED** [review]
LGTM!

File: sdk/storage/azblob/container/client.go [hunk 3]
```diff
@@ -160,6 +160,23 @@ func (c *Client) Delete(ctx context.Context, options *DeleteOptions) (DeleteResp
 	return resp, err
 }
 
+// Restore operation restore the contents and properties of a soft deleted container to a specified container.
+// For more information, see https://docs.microsoft.com/en-us/rest/api/storageservices/restore-container.
+func (c *Client) Restore(ctx context.Context, options *RestoreOptions) (RestoreResponse, error) {
```
Comment (nakulkar-msft) [inline sdk/storage/azblob/container/client.go:165]: Container Name and version are required parameters to restore a container. Both of these should be in the func signature. Options can be an empty struct for now. Else the using function with default value will always fail. 

Should we check if this API is needed at container level or just at service level?

File: sdk/storage/azblob/service/client.go [hunk 4]
```diff
@@ -123,6 +123,14 @@ func (s *Client) DeleteContainer(ctx context.Context, containerName string, opti
 	return containerDeleteResp, err
 }
 
+// RestoreContainer restores soft-deleted container
+// Operation will only be successful if used within the specified number of days set in the delete retention policy
+func (s *Client) RestoreContainer(ctx context.Context, deletedContainerName string, options *RestoreContainerOptions) (RestoreContainerResponse, error) {
```
Comment (nakulkar-msft) [inline sdk/storage/azblob/service/client.go:128]: Here as well. Required parameters should be in the signature.

**Review by nakulkar-msft — CHANGES_REQUESTED** [review]

File: sdk/storage/azblob/container/client.go (hunk 3 above)
Comment (jhendrixMSFT) [reply sdk/storage/azblob/container/client.go:165]: This is the `container` client, we already have the container name (see the following lines where we obtain it).

Comment (souravgupta-msft) [reply sdk/storage/azblob/container/client.go:165]: @jhendrixMSFT, should we add the deleted container version as a parameter of the function (this parameter is mandatory for restoring a container). We will also have the options as a parameter which will be empty for now. What do you think?
@nakulkar-msft, this is what you are referring to, right?

Comment (jhendrixMSFT) [reply sdk/storage/azblob/container/client.go:165]: If version is required for the API to work, then yeah it should be a required parameter and the options type empty.

Comment (souravgupta-msft) [reply sdk/storage/azblob/container/client.go]: Added deleted container version as a parameter and also keeping empty options struct

File: sdk/storage/azblob/service/client.go (hunk 4 above)
Comment (souravgupta-msft) [reply sdk/storage/azblob/service/client.go]: Added

**Review by vibhansa-msft — APPROVED** [review]

**Review by nakulkar-msft — APPROVED** [review]

**Pull request merged**